### ISSUE TITLE
Temp directory not found when plotting

### ISSUE THREAD

**Issue by PhilFreeman** [Calysto/octave_kernel#174]
Just started learning Notebooks today so maybe this has nothing to do with octave_kernel, but it seems to. I made a simple notebook with Jupyter Lab. Octave commands work fine and the system can evaluate and return results. However, plotting results in the following error

```
Inline plot failed, consider trying another graphics toolkit
warning: print.m: Ghostscript binary is not available.  Only eps output is possible
error: print: directory /tmp/tmp1_66g9ve does not exist
error: called from
    _make_figures>safe_print at line 125 column 7
    _make_figures at line 49 column 13
```

Checking the Octave history, I can see the call
`_make_figures("/tmp/tmp1_66g9ve", "png", "Figure", -1, -1, 0, 0)`

The warning is confusing because Ghostscript is installed and on the PATH. 
```
(base) phil@Office:~$ which gs
/usr/bin/gs
```
But this is clearly an Octave warning, so maybe I need to configure Ghostscript with Octave? I can plot directly from Octave without issues.

Details:
Ubuntu 18.04.4 LTS
Octave 5.2.0
Jupyter Lab 1.2.6

**Comment by blink1073** [comment]
Hi @PhilFreeman, the difference is that in the Octave kernel we are using `print` to get the plot output, which is not as well supported in Octave as the live figure windows.  If you try the following in Octave do you get a similar error?  `plot (1:10); print foo.png`

See also https://savannah.gnu.org/bugs/?56193

**Comment by PhilFreeman** [comment]
Yes, same error. I'll see if I can figure out a workaround.

**Comment by PhilFreeman** [comment]
Actually, it's not the same. I still get the warning (only on the first call), but the figure does print to the local working directory.
```
surf(peaks); print('peaks.png')
warning: print.m: Ghostscript binary is not available.  Only eps output is possible
warning: called from
    __print_parse_opts__ at line 487 column 7
    print at line 416 column 8
```
subsequent calls to print do not spawn the warning.
However, if I call ```print('tmp/peaks.png')``` then I get the error
```
error: print: directory tmp does not exist
error: called from
    print at line 422 column 7
```
Which indicates that Octave won't create the temporary directory for the figures. So, it seems that octave_kernel is requesting the files be saved to a directory that hasn't been created. Can I make a local figure directory and point octave_kernel to use that specifically?

**Comment by blink1073** [comment]
That is odd, the docs for [`mkdtemp`](https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp) say that it makes a directory, and it does for me locally using Python 3.8.1.

What version of Python are you using?

**Comment by PhilFreeman** [comment]
Python is 3.7.6. It looks like I have a typo in my test above. I wrote ```mktemp```, not ```mkdtemp``` (which it's really odd that I got anything returned, but I did). When I use ```mkdtemp```, indeed, it does create a temporary directory. Odd.

I'm going to spend some time setting up a development environment and try to learn to step through the code in a debugger. Maybe I can figure this out on my side. If I find something, I'll submit a pull request.

**Comment by blink1073** [comment]
You may also try running _make_figures directly from Octave.

**Comment by PhilFreeman** [comment]
I did. From the previous comment

> I cloned the repository, and then from within Octave with octave_kernel as the working directory I tried the following:
>
> ```>>>surf(peaks);```
> ```>>>_make_figures('.','png','Figure',-1,-1,0,0);```
>
> And that worked fine. The file Figure001.png was written to the local working directory.

**Comment by blink1073** [comment]
Ah, my mistake, too much context switching this morning.   I'm curious, if you comment out these lines locally, does it display the figure?   I wonder if we're being to aggressive in the error handling:

https://github.com/Calysto/octave_kernel/blob/6942177335988b04fa8c33d328056380e94dd7dc/octave_kernel/kernel.py#L282-L287

**Comment by PhilFreeman** [comment]
I'll try that. How do I get Jupyter to use my clone of the repo rather than the "installed" version of octave_kernel? I tried to do a ```make all``` from ```octave_kernel/```. That ran without error, but it doesn't seem that Jupyter is using the version I just made. Is there another step needed?

**Comment by blink1073** [comment]
`pip install -e .` from the repo root should do it.

**Comment by PhilFreeman** [comment]
So.... I screwed up. :) Commenting the lines only caused it to fail silently. Since the error strongly pointed towards Octave not finding the temp directory, I just hardcoded plot_dir to '.' and tried that. The error message changed so that seemed like progress. Then, octave_kernel dutifully cleaned up the temporary directory. I was running Jupyter from my home directory. Yeah. Lost everything.

I'll be spending the next evening trying to recover what I can. Fortunately, I don't really have anything critical/valuable on the machine I was using. Lesson definitely learned to always use a container, even if you think your just tinkering with a simple module. Once I get back up, I'll see if I can make progress again and post any updates.

The new error message did make it seem like it's an Octave problem with Ghostscript. Confusing why Octave can't find gs, but I'll sort through that later.

**Comment by blink1073** [comment]
Holy moly, I'm so sorry @PhilFreeman, that is a nightmare.

**Comment by PhilFreeman** [comment]
The ```pip install -e .``` doesn't seem to work for me. It runs, and claims success, but when I launch jupyter lab the Octave kernel option isn't available. My installation is Anaconda. Does that make a difference?

**Comment by blink1073** [comment]
Ah, right, you need `python -m octave_kernel install`.

**Comment by PhilFreeman** [comment]
OK, I finally figured it out. On Ubuntu Octave 5.2.0 is installed as a Snap application, so it's containerized. Ghostscript is not in that container, which screws up the ```print()``` command in CLI mode (but not in GUI mode, for some reason). I dropped back to 4.2.2, which is an apt version. Octave_kerenel works as designed and the inline graphics show up perfectly. I'll close this issue and see if I can find a way to let the team that packaged the snap version of Octave know that ghostscript is still really helpful.

**Comment by dsblank** [comment]
Thanks @PhilFreeman !

**Comment by slel** [comment]
> [...] and see if I can find a way to let the team that packaged
> the snap version of Octave know that ghostscript is still really helpful.

Open a ticket in the issue tracker for the "octave-snap" repo?

- https://github.com/octave-snap/octave-snap

**Comment by PhilFreeman** [comment]
I did a little more work on this. I'm not reopening the issue, but the kernel could be modified to work with the snap installation. I hacked this together by directly editing kernel.py in my pkgs. I haven't been able to make a working version of the repo so I can properly write this for a pull request. I'll outline the solution here.

With the Octave 5.2.0 Snap installation (and I assume any 5.x installation), switching the executable from ```octave-cli``` to simply ```octave``` works. The user can set the environment variable ```OCTAVE_EXECUTABLE = octave``` to force the right executable. The test for cli can be skipped if the version is 5 or greater. I simply commented out the CLI check. It wasn't a problem.

The next problem is that the Snap container can access the users home directory, but not the system temp directory. The workaround for the plot directory to be in the user's home path. It can't be hidden. I simply made a directory ```/plots``` and hard coded ```plot_dir``` in the kernel for testing. It worked fine. Really, we should add another plotting option for the user to specify a directory. This can be set from the .octave_kernel_config.py

If the user specifies a directory, then the deletion of the temp directory should be set to False. Ideally, the files _inside_ the directory should be cleaned up, but the directory itself should stay. Alternatively, the ```mkdtmp()``` call could use the user specified plot directory as its root.

Bottom line, a workaround can be made pretty easily. Or Octave Snap can add Ghostscript to its package. The work around seems pretty straight forward.

**Comment by blink1073** [comment]
Thanks @PhilFreeman!  I'll leave this open in case someone wants to implement the fix.  I am not an Ubunu user.

**Comment by PhilFreeman** [comment]
Issue fixed in version 0.32.0.
